backend/app/Main.hs

```diff
@@ -47,16 +44,17 @@ main :: IO ()
 main = do
   createDirectoryIfMissing True . takeDirectory $ databaseFile
 
-  -- TODO Fix up debug logging
-  dbPool <- runStdoutLoggingT $ createSqlitePoolWithConfig (toText databaseFile) defaultConnectionPoolConfig
+  let logFilter = filterInfo
+  logger <- stdoutLogger
+  dbPool <- runAppLogger logFilter logger $ createSqlitePoolWithConfig (toText databaseFile) defaultConnectionPoolConfig
   redisPool <- connect redisConfig
-  timeCache <- newTimeCache simpleTimeFormat
-  (logger, _) <- newTimedFastLogger timeCache logType
 
   let env = Env {dbPool, redisPool, logger}
 
   -- TODO Disable/handle auto-migration
-  runSqlPool (runMigration migrateAll) dbPool
+  migrations <- runSqlPool (runMigrationQuiet migrateAll) dbPool
+  unless (null migrations) (log logger LevelWarn "Database schema changed. Running migrations.")
+  mapM_ (log logger LevelInfo . toLogStr) migrations
```

Comment:
Yeah, totes. The default behavior of `runMigration` is to print the executed SQL commands to stderr. What `runMigrationQuiet` does is to return them as a `[Text]` instead of printing. So yeah, basically just trying to reclaim some control over how this will display. Output on stderr will just quietly disappear when this is running on a server, so this lets us log (or not log) in our standardized format.

`mapM` is quite useful. This line does exactly what you say. It's important (for Haskell learning) to understand the general case of what's being done here: `mapM` takes as its input: 
1. a function from `Monad m => a -> m b`
2. a bunch of `a`s: `t a`

If we did a normal `map` operation, we would end up with `t (m b)`, which could very well not be what we want (*maybe* we want a bunch of monads, but probably not). In the example of the code above, I'd have `[IO ()]` which sucks. So what `mapM` gives us is the ability to run all those monadic actions (the `a -> m b` applied to all the inputs), but then collect the results and wrap them all in the monad. So:

`map`, on `t a`, given `a -> m b`, yields `t (m b)`
`mapM`, on `t a`, given `a -> m b`, yields `m (t b)`

Much better! Now, in the case of our logging stuff, we end up with `IO [()]`. Nice to have the results collected under one monad - but these results are kinda stupid. We really only wanted to run the actions for their side-effect. So that's what `mapM_` is for. The only difference between `mapM` and `mapM_` is that `mapM_` discards the results - in other words, it always returns `m ()`. It's not just IO where this is useful, We can do this *even if* our monadic actions *do* return a result. Just depends on what you want :shrug: 

One more thing to note is that `mapM` is a monadic-specific version of the more general function `traverse` (which works on Applicatives). The type signature is nearly identical, differing only in the type constraint.

`traverse :: Applicative f => (a -> f b) -> t a -> f (t b)`
`mapM :: Monad m => (a -> m b) -> t a -> f (m b)`

Reply:
As a good rule of thumb:

If you end up with "a collection of results wrapped in monads", but what you wanted was "a monad wrapping a collection of results", you probably want to reach for `traverse` / `mapM`

Reply:
This breakdown is helpful! I didn't think to compare `mapM` to `mapM_`, that clarifies what the underscore is distinguishing. And nice to have `traverse` and `mapM` in the back pocket